genePred/genePred.go

```
@@ -0,0 +1,173 @@
+
+//Write writes a slice of GenePred structs with a specified number of fields to a specified filename.
+func Write(filename string, records []*GenePred) {
+	file := fileio.EasyCreate(filename)
+	defer file.Close()
+
+	WriteToFileHandle(file, records)
+}
+
+func Read(filename string) []*GenePred {
+	var line string
+	var answer []*GenePred
+	var doneReading = false
+
+	file := fileio.EasyOpen(filename)
+	defer file.Close()
+
+	for line, doneReading = fileio.EasyNextRealLine(file); !doneReading; line, doneReading = fileio.EasyNextRealLine(file) {
+		current := processGenePredLine(line)
+		answer = append(answer, current)
+	}
+	return answer
+}
+
+func processGenePredLine(line string) *GenePred {
+	current := GenePred{}
+
+	words := strings.Split(line, "\t")
+	current.Id = words[0]
+	current.Symbol = words[0]
+	current.Chrom = words[1]
+	if words[2] == "+" {
+		current.Strand = '+'
+	} else if words[2] == "." {
+		current.Strand = '.'
+	} else if words[2] == "-" {
+		current.Strand = '-'
+	} else {
+		log.Fatal("no strand specified")
+	}
+	current.TxStart = common.StringToInt(words[3])
+	current.TxEnd = common.StringToInt(words[4])
+	current.CdsStart = common.StringToInt(words[5])
+	current.CdsEnd = common.StringToInt(words[6])
+	current.ExonStarts = StringToIntSlice(words[7])
+	current.ExonEnds = StringToIntSlice(words[8])
+	current.ExonFrames = CalcExonFrame(&current)
+	current.Score = 0
+	exonNumber := len(current.ExonStarts)
+
+	if exonNumber != len(current.ExonStarts) {
+		//DEBUG: log.Print(exonNumber)
+		log.Fatal("exon number does not equal number of start coordinates")
+	}
+
+	if len(current.ExonStarts) != len(current.ExonEnds) {
+		log.Fatal("there are not the same number of exon start positions as exon end positions")
+	}
+
+	return &current
+}
+
+func StringToIntSlice(text string) []int {
+	values := strings.Split(text, ",")
+	var answer = make([]int, len(values)-1)
+
+	for i := 0; i < len(values)-1; i++ {
+		answer[i] = common.StringToInt(values[i])
+	}
```

One thing we could consider doing is to check the final element in the slice (`values[len(values)-1]`) if it equals 0 and log an error if it doesn't.

---

Would this be the same as checking if the last item of values is equal to the empty string?

---

I was going to ask the same thing, we can talk about it in code review I think so we are all on the same page

---

I think if you call `common.StringToInt(values[i])` on the empty string, you get 0 back, but I have to check

---

We should probably change common.StringToInt to exit with an error if it gets the empty string.  We could add a TODO about that (or a pull request to fix it) if it is silently returning zeros.  Maybe we could use strings.HasSuffix here to make sure that "text" ends with a comma?

---

I think checking the suffix for a comma is a great idea! Don't quote me on the common.StringToInt(), I'll have to double check, I am just going off what happened last time when I programmed this and a TODO to double check common.StringToInt would be great, but as for this check in, i think adding something on line 109 to have a check like: `if values[len(values)-1] != "" { log.Fatalf("Error: Last element in values should be equal to the empty string")}` is something I would do.

---

I added 2 checks on lines 89-98 for the suffix check